tasks/check_legacy_compilers.rb

```diff
@@ -0,0 +1,67 @@
+#!/opt/puppetlabs/puppet/bin/ruby
+# frozen_string_literal: true
+
+require 'json'
+require 'uri'
+require 'net/http'
+require 'puppet'
+
+# CheckLegacyCompilers task class
+class CheckLegacyCompilers
+  def initialize(params)
+    @nodes = params['legacy_compilers'].split(',') if params['legacy_compilers'].is_a?(String)
+  end
+
+  def execute!
+    pinned_nodes = []
+    @nodes.each do |node|
+      node_classification = get_node_classification(node)
+      pinned = false
+      node_classification['groups'].each do |group|
+        if group['name'] == 'PE Master'
+          pinned_nodes << node
+          pinned = true
+        end
+      end
+      next if pinned
+      next unless node_classification.key?('parameters')
+      next unless node_classification['parameters'].key?('pe_master')
+      if node_classification['parameters']['pe_master']
+        pinned_nodes << node
+      end
+    end
+
+    return unless !pinned_nodes.empty?
+    puts 'The following legacy compilers are classified as Puppet primary:'
+    puts pinned_nodes.join(', ')
+    puts 'You will not be able to upgrade if you dont remediate this.'
```

Comment:
The phrase "remediate this" is possibly too vague - however, we don't want to recommend a specific action as the required remedial action depends on customer workflow. 
We could use something that is a little clearer but still high level. For example:
"You will not be able to upgrade until these compilers are no longer classified as Puppet primary."